flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java

```diff
@@ -0,0 +1,118 @@
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Serializer class for {@link AsyncSinkWriter} state.
+ *
+ * @param <RequestEntryT> Writer Request Entry type
+ */
+@Internal
+public abstract class AsyncSinkWriterStateSerializer<RequestEntryT extends Serializable>
+        implements SimpleVersionedSerializer<BufferedRequestState<RequestEntryT>> {
+    private static final long DATA_IDENTIFIER = -1;
+
+    /**
+     * Serializes state in form of
+     * [DATA_IDENTIFIER,NUM_OF_ELEMENTS,SIZE1,REQUEST1,SIZE2,REQUEST2....].
+     */
+    @Override
+    public byte[] serialize(BufferedRequestState<RequestEntryT> obj) throws IOException {
+        Collection<RequestEntryWrapper<RequestEntryT>> bufferState =
+                obj.getBufferedRequestEntries();
+        try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                final DataOutputStream out = new DataOutputStream(baos)) {
+            out.writeLong(DATA_IDENTIFIER);
+            out.writeInt(bufferState.size());
+            for (RequestEntryWrapper<RequestEntryT> wrapper : bufferState) {
+                out.writeLong(wrapper.getSize());
+                serializeRequestToStream(wrapper.getRequestEntry(), out);
+            }
+            return baos.toByteArray();
+        }
+    }
+
+    @Override
+    public BufferedRequestState<RequestEntryT> deserialize(int version, byte[] serialized)
+            throws IOException {
+        try (final ByteArrayInputStream bais = new ByteArrayInputStream(serialized);
+                final DataInputStream in = new DataInputStream(bais)) {
+            validateIdentifier(in);
+            int size = readInt(in);
+            List<RequestEntryWrapper<RequestEntryT>> serializedState = new ArrayList<>();
+            for (int i = 0; i < size; i++) {
+                long requestSize = readLong(in);
+                RequestEntryT request = deserializeRequestFromStream(requestSize, in);
+                serializedState.add(new RequestEntryWrapper<>(request, requestSize));
+            }
+
+            return new BufferedRequestState<>(serializedState);
+        }
+    }
+
+    protected abstract void serializeRequestToStream(RequestEntryT request, DataOutputStream out)
+            throws IOException;
+
+    protected abstract RequestEntryT deserializeRequestFromStream(
+            long requestSize, DataInputStream in) throws IOException;
+
+    private void validateIdentifier(DataInputStream in) throws IOException {
+        if (readLong(in) != DATA_IDENTIFIER) {
+            throw new IllegalStateException("Corrupted data to deserialize");
+        }
+    }
+
+    protected Long readLong(DataInputStream in) throws IOException {
+        try {
+            return in.readLong();
+        } catch (IndexOutOfBoundsException | EOFException e) {
+            throw new IllegalStateException("Incomplete serialized state");
```

Comment:
If we are going to rethrow we should wrap `e`. However, I would be inclined to just remove these methods `readLong`/`readInt` directly and let the exceptions bubble up. This is similar to other implementations. Generally speaking it is bad practise to catch `RuntimeException`